test/parallel_api/algorithm/alg.sorting/sort_by_key_common.h

```diff
@@ -0,0 +1,289 @@
+    auto has_duplicates = [&histogram](const auto& pair) {return histogram[std::get<0>(pair)] > 1; };
+    auto new_last = std::remove_if(first, first + n, has_duplicates);
+    return std::distance(first, new_last);
+}
+
+template<typename KeyIt, typename ValIt, typename Size>
+void
+generate_data(KeyIt keys_begin, ValIt vals_begin, Size n, std::uint32_t seed)
+{
+    using KeyT = typename std::iterator_traits<KeyIt>::value_type;
+    using ValT = typename std::iterator_traits<ValIt>::value_type;
+    if constexpr (std::is_same_v<ValT, Particle>)
+    {
+        static_assert(std::is_same_v<KeyT, Particle::energy_type>);
+        std::default_random_engine gen{seed};
+        std::uniform_real_distribution<float> mass_dist(0.0, 1000.0);
+        std::uniform_real_distribution<float> velocity_dist(0.0, 1.0);
+        std::uniform_real_distribution<float> coord_dist(0.0, 1.0);
+        std::generate(vals_begin, vals_begin + n, [&gen, &mass_dist, &velocity_dist, &coord_dist]() {
+            return Particle{mass_dist(gen), velocity_dist(gen), {coord_dist(gen), coord_dist(gen), coord_dist(gen)}};
+        });
+        std::transform(vals_begin, vals_begin + n, keys_begin, [](const auto& particle) { return particle.energy(); });
+    }
+    else
+    {
+        TestUtils::generate_arithmetic_data(keys_begin, n, seed);
+        TestUtils::generate_arithmetic_data(vals_begin, n, seed + 1);
+        // avoid having value duplicates at the same locations as key duplicates for a more robust stability check
+        std::shuffle(vals_begin, vals_begin + n, std::default_random_engine(seed));
+    }
+}
+
+template<typename Policy, typename KeyIt, typename ValIt, typename Size, typename Compare>
+void
+call_sort(Policy&& policy, KeyIt keys_begin, ValIt vals_begin, Size n, Compare compare, StableSort)
+{
+    if constexpr (std::is_same_v<Compare, DefaultComparator>)
+        oneapi::dpl::stable_sort_by_key(policy, keys_begin, keys_begin + n, vals_begin);
+    else
+        oneapi::dpl::stable_sort_by_key(policy, keys_begin, keys_begin + n, vals_begin, compare);
+}
+
+template<typename Policy, typename KeyIt, typename ValIt, typename Size, typename Compare>
+void
+call_sort(Policy&& policy, KeyIt keys_begin, ValIt vals_begin, Size n, Compare compare, UnstableSort)
+{
+    if constexpr (std::is_same_v<Compare, DefaultComparator>)
+        oneapi::dpl::sort_by_key(policy, keys_begin, keys_begin + n, vals_begin);
+    else
+        oneapi::dpl::sort_by_key(policy, keys_begin, keys_begin + n, vals_begin, compare);
+}
+
+template<typename KeyIt, typename ValIt, typename Size, typename Compare>
+void
+call_reference_sort(KeyIt ref_keys_begin, ValIt ref_vals_begin, Size n, Compare)
+{
+    static_assert(std::is_same_v<Compare, DefaultComparator> || std::is_same_v<Compare, GreaterComparator>);
+    auto first = oneapi::dpl::make_zip_iterator(ref_keys_begin, ref_vals_begin);
+    if constexpr (std::is_same_v<Compare, DefaultComparator>)
```

Comment:
I dont understand why we need 138-149 logic.
lets call sort for testing with Comparator type which is passed by a test case.

Reply:
`dpl::sort_by_key` is checked by `std::stable_sort`. Their comparators have different semantics.

For example: 
`std::stable_sort(..., [](const auto& lhs, const auto& rhs) { return std::get<0>(lhs) > std::get<0>(rhs)})` checks 
`dpl::sort_by_key(..., [](const T& lhs, const T& rhs) { return lhs > rhs; })`.

That is why the same comparator cannot be used.

Reply:
`comp(std::get<0>(lhs), std::get<0>(rhs))`  - the same comparator

Reply:
I've reused that idea in https://github.com/oneapi-src/oneDPL/pull/1692#discussion_r1716714607, but still there is `if constexpr` branching.
I've created https://github.com/oneapi-src/oneDPL/issues/1780 for further refactoring.